metric_providers/disk/io/cgroup/container/source.c

```diff
@@ -0,0 +1,189 @@
+static int user_id = 0;
+static unsigned int msleep_time=1000;
+
+
+static disk_io_t get_disk_cgroup(char* filename) {
+    long int rbytes = -1;
+    long int rbytes_sum = 0;
+    long int wbytes = -1;
+    long int wbytes_sum = 0;
+
+    FILE * fd = fopen(filename, "r");
+    if ( fd == NULL) {
+        fprintf(stderr, "Error - Could not open path for reading: %s. Maybe the container is not running anymore? Are you using --rootless mode? Errno: %d\n", filename, errno);
+        exit(1);
+    }
+
+    while (fscanf(fd, "%*u:%*u rbytes=%ld wbytes=%ld rios=%*u wios=%*u dbytes=%*u dios=%*u", &rbytes, &wbytes) == 2) {
+        rbytes_sum += rbytes;
+        wbytes_sum += wbytes;
+    }
+    ;
+
+    fclose(fd);
+
+    if(rbytes < 0 || wbytes < 0) {
+        fprintf(stderr, "Error - io.stat could not be read or was < 0.");
+        exit(1);
+    }
+
+    disk_io_t disk_io;
+    disk_io.rbytes = rbytes_sum;
+    disk_io.wbytes = wbytes_sum;
+    return disk_io;
+}
+
+static int output_stats(container_t *containers, int length) {
+
+    struct timeval now;
+    int i;
+
+    gettimeofday(&now, NULL);
+    for(i=0; i<length; i++) {
+        disk_io_t disk_io = get_disk_cgroup(containers[i].path);
+        printf("%ld%06ld %ld %s\n", now.tv_sec, now.tv_usec, disk_io.rbytes+disk_io.wbytes, containers[i].id);
+    }
+    usleep(msleep_time*1000);
+
+    return 1;
+}
+
+static int parse_containers(container_t** containers, char* containers_string, int rootless_mode) {
+    if(containers_string == NULL) {
+        fprintf(stderr, "Please supply at least one container id with -s XXXX\n");
+        exit(1);
+    }
+
+    *containers = malloc(sizeof(container_t));
+    char *id = strtok(containers_string,",");
+    int length = 0;
+
+    for (; id != NULL; id = strtok(NULL, ",")) {
+        //printf("Token: %s\n", id);
+        length++;
+        *containers = realloc(*containers, length * sizeof(container_t));
+        (*containers)[length-1].id = id;
+        if(rootless_mode) {
+            sprintf((*containers)[length-1].path,
+                "/sys/fs/cgroup/user.slice/user-%d.slice/user@%d.service/user.slice/docker-%s.scope/io.stat",
+                user_id, user_id, id);
+        } else {
+            sprintf((*containers)[length-1].path,
+                "/sys/fs/cgroup/system.slice/docker-%s.scope/io.stat",
+                id);
+        }
+    }
+
+    if(length == 0) {
+        fprintf(stderr, "Please supply at least one container id with -s XXXX\n");
+        exit(1);
+    }
+    return length;
+}
+
+static int check_system(int rootless_mode) {
+    const char* check_path;
+
+    if(rootless_mode) {
+        check_path = "/sys/fs/cgroup/user.slice/io.stat";
+    } else {
+        check_path = "/sys/fs/cgroup/system.slice/io.stat";
+    }
+
+    FILE* fd = NULL;
+    fd = fopen(check_path, "r");
+
+    if (fd == NULL) {
+        fprintf(stderr, "Couldn't open io.stat file at %s\n", check_path);
+        exit(127);
```

Comment:
Why `127`?

https://stackoverflow.com/questions/1763156/127-return-code-from

you use `1` everywhere else

Reply:
normalized to 1 where possible in all files